Signal/src/Profiles/OWSProfileManager.m

```diff
@@ -225,150 +219,170 @@ - (UserProfile *)getOrBuildUserProfileForRecipientId:(NSString *)recipientId
     return instance;
 }
 
-// All writes to user profiles should occur on the main thread.
 - (void)saveUserProfile:(UserProfile *)userProfile
 {
-    OWSAssert([NSThread isMainThread]);
     OWSAssert(userProfile);
 
-    // Make sure to save on the local db connection for consistency.
-    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
-        [userProfile saveWithTransaction:transaction];
-    }];
+    @synchronized(self)
+    {
+        // Make sure to save on the local db connection for consistency.
+        [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+            [userProfile saveWithTransaction:transaction];
+        }];
 
-    if (userProfile == self.localUserProfile) {
-        [[NSNotificationCenter defaultCenter] postNotificationName:kNSNotificationName_LocalProfileDidChange
-                                                            object:nil
-                                                          userInfo:nil];
-    } else {
-        [[NSNotificationCenter defaultCenter] postNotificationName:kNSNotificationName_OtherUsersProfileDidChange
-                                                            object:nil
-                                                          userInfo:nil];
+        BOOL isLocalUserProfile = userProfile == self.localUserProfile;
+
+        dispatch_async(dispatch_get_main_queue(), ^{
+            if (isLocalUserProfile) {
+                [[NSNotificationCenter defaultCenter] postNotificationName:kNSNotificationName_LocalProfileDidChange
+                                                                    object:nil
+                                                                  userInfo:nil];
+            } else {
+                [[NSNotificationCenter defaultCenter]
+                    postNotificationName:kNSNotificationName_OtherUsersProfileDidChange
```

Comment:
seems likely we'll want to include the recipientId or the userProfile in the notification.